tango/workspaces/remote_workspace.py

```diff
@@ -0,0 +1,344 @@
+            return result
+
+        step_info = self.step_info(step)
+        if step_info.state != StepState.RUNNING:
+            raise StepStateError(step, step_info.state)
+
+        # Update step info and save step execution metadata.
+        # This needs to be done *before* adding the result to the cache, since adding
+        # the result to the cache will commit the step dataset, making it immutable.
+        step_info.end_time = utc_now_datetime()
+        step_info.result_location = self.client.url(self.Constants.step_dataset_name(step))
+        self._update_step_info(step_info)
+
+        self.cache[step] = result
+        if hasattr(result, "__next__"):
+            assert isinstance(result, Iterator)
+            # Caching the iterator will consume it, so we write it to the cache and then read from the cache
+            # for the return value.
+            result = self.cache[step]
+
+        self.locks.pop(step).release()
+
+        return result
+
+    def step_failed(self, step: Step, e: BaseException) -> None:
+        # We don't do anything with uncacheable steps.
+        if not step.cache_results:
+            return
+
+        try:
+            step_info = self.step_info(step)
+            if step_info.state != StepState.RUNNING:
+                raise StepStateError(step, step_info.state)
+            step_info.end_time = utc_now_datetime()
+            step_info.error = exception_to_string(e)
+            self._update_step_info(step_info)
+        finally:
+            self.locks.pop(step).release()
+
+    def register_run(self, targets: Iterable[Step], name: Optional[str] = None) -> Run:
+        import concurrent.futures
+
+        all_steps = set(targets)
+        for step in targets:
+            all_steps |= step.recursive_dependencies
+
+        # Create a remote dataset that represents this run. The dataset which just contain
+        # a JSON file that maps step names to step unique IDs.
+        run_dataset: RemoteDataset
+        if name is None:
+            # Find a unique name to use.
+            while True:
+                name = petname.generate() + str(random.randint(0, 100))
+                try:
+                    run_dataset = self.client.create(
+                        self.Constants.run_dataset_name(cast(str, name)), commit=False
+                    )
+                except RemoteDatasetConflict:
+                    continue
+                else:
+                    break
+        else:
+            try:
+                run_dataset = self.client.create(
+                    self.Constants.run_dataset_name(name), commit=False
+                )
+            except RemoteDatasetConflict:
+                raise ValueError(f"Run name '{name}' is already in use")
+
+        steps: Dict[str, StepInfo] = {}
+        run_data: Dict[str, str] = {}
+
+        # Collect step info.
+        with concurrent.futures.ThreadPoolExecutor(
+            thread_name_prefix="GCSWorkspace.register_run()-"
```

Comment:
Maybe something more generic?
```suggestion
            thread_name_prefix="RemoteWorkspace.register_run()-"
```